validate codes list for saved codes tab

sb-prateekkeerthi opened the pull request:
## Terminology Service PR

Jira Ticket: [MAT-6856](https://jira.cms.gov/browse/MAT-6856)
(Optional) Related Tickets:

### Summary
validate codes list for saved codes tab

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

sb-prateekkeerthi committed:
validate codes list for saved codes tab

sb-prateekkeerthi committed:
added test cases

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java:
```diff
@@ -85,4 +86,13 @@ public ResponseEntity<Code> getCode(
         .body(fhirTerminologyService.retrieveCode(code, codeSystem, version, user.getApiKey()));
   }
+
+  @PostMapping(path = "/codesList", produces = MediaType.APPLICATION_JSON_VALUE)
```
Preferred: `/codes` to indicate it is a collection of `Code` resources.

sb-prateekkeerthi replied:
thanks for the comment I have changed it to /codes

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/controller/VsacFhirTerminologyController.java:
```diff
@@ -85,4 +86,13 @@ public ResponseEntity<Code> getCode(
+
+  @PostMapping(path = "/codesList", produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<List<Code>> getCodesList(
+      @RequestBody() List<Map<String, String>> codeList, Principal principal) {
```
Suggestion: Rename `codeList` to better explain what the Strings are in the Map. For instance, if they are the code and code system, then `codeAndCodeSystems` perhaps. 

I got slightly confused by the method named `getCodesList` taking an argument called codeList, since that sounds like the expected return object.

sb-prateekkeerthi replied:
thanks for the comment I have changed it to getCodesAndCodeSystems

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
+                      : null;
+
+              if (mappedVersion != null) {
```
```suggestion
             if (mappedVersion.isPresent()) {
```
Recommend `isPresent` since `mappedVersion` is an `Optional`

sb-prateekkeerthi replied:
thank you. I have made the changes

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+              String oid = code.get("oid") != null ? code.get("oid").replaceAll("'|'", "") : null;
+
+              Optional<Map.Entry<String, String>> mappedVersion =
+                  oid != null
+                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
+                      : null;
```
Suggest moving the null check to the top of `mapToFhirVersion()`.

sb-prateekkeerthi replied:
thank you, I have made those changes

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+              String oid = code.get("oid") != null ? code.get("oid").replaceAll("'|'", "") : null;
+
+              Optional<Map.Entry<String, String>> mappedVersion =
+                  oid != null
+                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
+                      : null;
```
```suggestion
                  oid != null
                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
                      : Optional.empty();
```
Alternatively to moving the null check inside `mapToFhirVersion`, keep with the Optional pattern and return `Optional.empty()`.

sb-prateekkeerthi replied:
thank you, I have made those changes

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+    if (result.isPresent()) {
+      return result;
+    }
+    return null;
```
```suggestion
    return Optional.empty();
```
Stick with Optional pattern.

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+              .findFirst();
+    }
+
+    if (result.isPresent()) {
+      return result;
+    }
+    return null;
```
```suggestion
    return result;
```
Since `result` is an Optional, it can be returned directly and handled by the `mappedVersion.isPresent()` check.

sb-prateekkeerthi replied:
thank you. I have made the changes

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+
+  private Optional<Map.Entry<String, String>> mapToFhirVersion(
+      String version, String oid, List<CodeSystemEntry> codeSystemEntries) {
+
```
```suggestion
    if (oid == null) {
      return Optional.empty();
    }
```
Recommend moving the null check to here to avoid having to check for null before every usage of this method (which, so far, there's only 1).

sb-prateekkeerthi replied:
I moved it to here to make it more generic

sb-prateekkeerthi committed:
resolved the feedback changes

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -278,4 +277,104 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+              .findFirst();
+    }
+
+    if (result.isPresent()) {
+      return result;
+    }
+    return result;
```
```suggestion
    return result;
```
Can be simplified to the above.

sb-prateekkeerthi replied:
thank you, I just fixed it

jkotanchik-SB commented on src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java:
```diff
@@ -278,4 +277,104 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
+          Code.builder()
+              .name(codeName)
+              .codeSystem(codeSystemName)
+              .version(vsacVersion)
```
Should pass `fhirVersion`. `vsacService.getCodeStatus` compares `version` against the fhirVersion in the mapping doc to look up the SVS version (which is known at this point)
```suggestion
              .version(fhirVersion)
```

jkotanchik-SB replied:
Turns out, we needed both versions, so I added an `svsVersion` entry to the DTO.

jkotanchik-SB committed:
MAT-6856: Use Code System OID when performing VSAC SVS Version lookup instead of Code System Name.

The OID is less prone user-entry errors in the mapping doc and more consistent across services.

jkotanchik-SB committed:
MAT-6856: During Code Status check, skip version map lookup if the VSAC SVS Version is already known.

jkotanchik-SB committed:
MAT-6856: Collapse `if` block since caller will have to handle the returned Optional regardless.

jkotanchik-SB committed:
MAT-6856: Add VSAC SVS Version to DTO so that we can perform the lookup once and store it on the object.

jkotanchik-SB committed:
MAT-6856: formatter

jkotanchik-SB committed:
MAT-6856: Remove dead code

jkotanchik-SB committed:
MAT-6856: Remove non-specific try/catch to prevent hiding actual errors.

jkotanchik-SB committed:
MAT-6856: formatter

jkotanchik-SB committed:
MAT-6856: Called method would always return null for provided arguments. Simply statement to directly return null.

jkotanchik-SB committed:
MAT-6856: Remove redundant toString's, and use null safe String equals.
